### DISCUSSION TITLE
Understanding kafka-coordinator-heartbeat-thread

### DISCUSSION THREAD

**Discussion by mandeepjain** [spring-projects/spring-kafka#2446] [Q&A]
We were investigating Thread dump of one of the servers and found out there were multiple kafka heartbeat-threads in waiting state.

I wanted to understand why these threads are in waiting state.

<img width="1454" alt="Screenshot 2022-10-19 at 3 27 21 PM" src="https://user-images.githubusercontent.com/47387463/196659991-20080992-dec2-4e43-9201-83c13aaad618.png">

**Comment by Wzy19930507** [comment]
Source code in kafka client class `AbstractCoordinator#HeartbeatThread.run`, look like
```java
        @Override
        public void run() {
            while (true) {
                // ...
                } else if (!heartbeat.shouldHeartbeat(now)) {
                    // poll again after waiting for the retry backoff in case the heartbeat failed or the
                    // coordinator disconnected
                    AbstractCoordinator.this.wait(rebalanceConfig.retryBackoffMs);
                } else {
                    heartbeat.sentHeartbeat(now);
                    // ...
                }
            }
        }
```

1. a consumer start a HeartThread
2. kafka client call `Heartbeat.entHeartbeat` every period `heartbeatIntervalMs`

**Comment by artembilan** [comment]
I'm not sure why we have such a discussion here in Spring project.
It looks like something internal to Apache Kafka client by itself: better to ask the question in their community.
